### PR TITLE
Cpm 752 : [back] Functional validation

### PR DESCRIPTION
**Pull request by Beninho**
**Description (for Contributor and Core Developer)**

Functionnal validation for :
 - Target Attribute does not exist
 - Target Attribute is not identifier
 - Limit of 1 identifier generator
 - Code not empty

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by cemorin — APPROVED** [review]

File: components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/validators.en_US.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+validation:
+    create:
+        target_attribute_does_not_exist: "The {{code}} attribute code given as target does not exist"
+        target_attribute_is_not_an_identifier: "The {{code}} attribute code is {{type}} type and should be of type identifier"
+        identifier_limit_reached: "Limit of {{limit}} identifier generator is reached"
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/validators.en_US.yml:5]:
```suggestion
        identifier_limit_reached: "Limit of {{limit}} identifier generators is reached"
```

File: components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/validators.en_US.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+validation:
+    create:
+        target_attribute_does_not_exist: "The {{code}} attribute code given as target does not exist"
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Resources/translations/validators.en_US.yml:3]: Can you wrap every {{ parameter }} with double quotes in this file plz ?

File: components/identifier-generator/back/tests/.php_cd.php [hunk 3]
```diff
@@ -16,28 +16,46 @@
...
+            'Akeneo\Pim\Structure\Component\Query\PublicApi',
+            'Akeneo\Platform\Bundle\FrameworkBundle\Security\SecurityFacadeInterface',
+            'Akeneo\UserManagement\Bundle\Context\UserContext',
 
             'Symfony\Component\HttpKernel\Bundle\Bundle',
             'Symfony\Component\Config\FileLocator',
             'Symfony\Component\DependencyInjection\ContainerBuilder',
             'Symfony\Component\DependencyInjection\Loader\YamlFileLoader',
             'Symfony\Component\HttpKernel\DependencyInjection\Extension',
+            'Symfony\Component\Validator\Constraint',
```
Comment (pierallard) [inline components/identifier-generator/back/tests/.php_cd.php:49]: IMHO You can only add 'Symfony\Component\Validator\Constraint'
And 'Symfony\Component\HttpFoundation'.

File: components/identifier-generator/back/tests/Specification/Infrastructure/Repository/InMemoryIdentifierGeneratorRepositorySpec.php [hunk 4]
```diff
@@ -77,8 +74,25 @@ function it_can_retrieve_an_identifier_generator_with_its_code()
         $this->get('aabbcc')->shouldBeLike($identifierGenerator);
     }
 
-    function it_returns_null_if_identifier_generator_is_not_found()
+    public function it_returns_null_if_identifier_generator_is_not_found()
     {
         $this->get('unknown')->shouldReturn(null);
     }
+
+    public function it_can_count_identifer_generators()
```
Comment (pierallard) [inline components/identifier-generator/back/tests/Specification/Infrastructure/Repository/InMemoryIdentifierGeneratorRepositorySpec.php:82]:
```suggestion
    public function it_counts_identifier_generators()
```

File: components/identifier-generator/back/tests/Specification/Infrastructure/Validation/IdentifierGeneratorCreationLimitValidatorSpec.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+use Akeneo\Pim\Automation\IdentifierGenerator\Infrastructure\Validation\IdentifierGeneratorCreationLimit;
+use Akeneo\Pim\Automation\IdentifierGenerator\Infrastructure\Validation\IdentifierGeneratorCreationLimitValidator;
+use PhpSpec\ObjectBehavior;
+use Prophecy\Argument;
+use Symfony\Component\Validator\Constraints\NotBlank;
+use Symfony\Component\Validator\Context\ExecutionContext;
+
+class IdentifierGeneratorCreationLimitValidatorSpec extends ObjectBehavior
+{
+    public function let(InMemoryIdentifierGeneratorRepository $repository, ExecutionContext $context): void
```
Comment (pierallard) [inline components/identifier-generator/back/tests/Specification/Infrastructure/Validation/IdentifierGeneratorCreationLimitValidatorSpec.php:18]: Should we use an interface here instead ?

File: components/identifier-generator/back/tests/Specification/Infrastructure/Validation/TargetAttributeShouldBeAnIdentifierValidatorSpec.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            ->willReturn(new CreateGeneratorCommand('generatorCode', [], [], [], 'sku', '-'));
+
+        $context->buildViolation(
+            'validation.create.target_attribute_is_not_an_identifier',
+            ['{{code}}' => 'sku', '{{type}}' => 'pim_catalog_text']
+        )->shouldBeCalled();
+
+        $this->validate('sku', new TargetAttributeShouldBeAnIdentifier());
+    }
+
```
Comment (pierallard) [inline components/identifier-generator/back/tests/Specification/Infrastructure/Validation/TargetAttributeShouldBeAnIdentifierValidatorSpec.php:77]: DOUBLE EMPTY LINE :rage:

**Review by pierallard — APPROVED** [review]

**Pull request merged**